File: Elixir/0004.exs
defmodule Solution do
  @spec find_median_sorted_arrays(nums1 :: [integer], nums2 :: [integer]) :: float
  def find_median_sorted_arrays(nums1, nums2) do
    # Ensure nums1 is the smaller array to make the binary search more efficient
    if length(nums1) > length(nums2) do
      find_median_sorted_arrays(nums2, nums1)
    else
      binary_search(nums1, nums2, 0, length(nums1))
    end
  end

  defp binary_search(nums1, nums2, low, high) do
    m = length(nums1)
    n = length(nums2)

    if low <= high do
      i = div(low + high, 2)
      j = div(m + n + 1, 2) - i

      max_left_1 = if i == 0, do: -1.0e308, else: Enum.at(nums1, i - 1)
      min_right_1 = if i == m, do: 1.0e308, else: Enum.at(nums1, i)
      
      max_left_2 = if j == 0, do: -1.0e308, else: Enum.at(nums2, j - 1)
      min_right_2 = if j == n, do: 1.0e308, else: Enum.at(nums2, j)

      if max_left_1 <= min_right_2 and max_left_2 <= min_right_1 do
        # Found the correct partition
        if rem(m + n, 2) == 0 do
          (max(max_left_1, max_left_2) + min(min_right_1, min_right_2)) / 2
        else
          max(max_left_1, max_left_2)
        end
      else
        if max_left_1 > min_right_2 do
          binary_search(nums1, nums2, low, i - 1)
        else
          binary_search(nums1, nums2, i + 1, high)
        end
      end
    end
  end
end

File: Erlang/0004.erl
-spec find_median_sorted_arrays(Nums1 :: [integer()], Nums2 :: [integer()]) -> float().
find_median_sorted_arrays(Nums1, Nums2) ->
    case length(Nums1) > length(Nums2) of
        true -> find_median_sorted_arrays(Nums2, Nums1);
        false ->
            X = length(Nums1),
            Y = length(Nums2),
            find_median(Nums1, Nums2, X, Y, 0, X)
    end.

-spec find_median(Nums1 :: [integer()], Nums2 :: [integer()], X :: integer(), Y :: integer(), Low :: integer(), High :: integer()) -> float().
find_median(Nums1, Nums2, X, Y, Low, High) when Low =< High ->
    PartitionX = (Low + High) div 2,
    PartitionY = (X + Y + 1) div 2 - PartitionX,

    MaxLeftX = if PartitionX =:= 0 -> -1000000000; true -> lists:nth(PartitionX, Nums1) end,
    MinRightX = if PartitionX =:= X -> 1000000000; true -> lists:nth(PartitionX + 1, Nums1) end,

    MaxLeftY = if PartitionY =:= 0 -> -1000000000; true -> lists:nth(PartitionY, Nums2) end,
    MinRightY = if PartitionY =:= Y -> 1000000000; true -> lists:nth(PartitionY + 1, Nums2) end,

    case MaxLeftX =< MinRightY andalso MaxLeftY =< MinRightX of
        true ->
            if (X + Y) rem 2 =:= 0 ->
                (max(MaxLeftX, MaxLeftY) + min(MinRightX, MinRightY)) / 2.0;
            true ->
                max(MaxLeftX, MaxLeftY)
            end;
        false ->
            if MaxLeftX > MinRightY ->
                find_median(Nums1, Nums2, X, Y, Low, PartitionX - 1);
            true ->
                find_median(Nums1, Nums2, X, Y, PartitionX + 1, High)
            end
    end;

find_median(_, _, _, _, _, _) ->
    erlang:error(invalid_argument).
